### DISCUSSION TITLE
TD Ameritrade API JSON

### DISCUSSION THREAD

**Discussion by cyclone-isu2001** [omegastripes/VBA-JSON-parser#26] [Q&A]
Trying to apply this with a TD Ameritrade API call.  When the JSON gets returned and the script runs, its response is "JSON has no rows".  Any input or advice would be greatly appreciated...  The JSON seems to parse using an online parser such as: http://json.parser.online.fr/
[return string.txt](https://github.com/omegastripes/VBA-JSON-parser/files/7108277/return.string.txt)

**Comment by omegastripes** [comment]
@cyclone-isu2001 remove the code after `JSON.Parse` line in the example, and create code, which depends on your needs.

**Reply by cyclone-isu2001** [reply]
sorry, I am very new to this.. I assumed that portion of the code was taking the parsed JSON and writing it to the worksheet.  Based on your response, I assume the JSON is parsing correctly, and I just need to write the elements out to the worksheets?

**Reply by omegastripes** [reply]
The JSON response you provided is not a simple table data, rather complex and nested data structure. You have to decide how to split it into separate tables and output the tables one by one to the worksheets.

**Reply by cyclone-isu2001** [reply]
OK - thank you....  i think this got well over my head...  I'll keep studying..  thanks for your responses...

has the parser created the tables?  is everything I want returned as vJSON?  can you give me an example of how I see, or inspect what is returned in vJSON and I can go from there?

**Reply by omegastripes** [reply]
Do you have any idea of how the results tables should look like? Or maybe there are some examples anywhere?

**Reply by cyclone-isu2001** [reply]
Its a stock quote and options pricing for a various strike prices..

![Capture](https://user-images.githubusercontent.com/90066957/132064600-56c5bee2-762a-4f42-b844-e765c857b1cf.JPG)

**Reply by omegastripes** [reply]
> has the parser created the tables? is everything I want returned as vJSON?

As readme says, `JSON.Parse` just parses JSON string and returns a structure of VBA arrays and dictionary objects which represents JSON string. Your task is to make further processing of that structure creating VBA code which extracts the necessary data, creates tables, outputs it, etc.. All depends on your needs.

**Reply by omegastripes** [reply]
> how I see, or inspect what is returned in vJSON and I can go from there?

There is a structure of JSON data you already displayed above - here you can inspect it and decide which parts to extract from it.

**Reply by cyclone-isu2001** [reply]
how is the data stored in vJSON?  I guess that's what I'm not understanding...    if its just another string what's the purpose of the parser?  i can just write a parser myself...  I thought it found the headings... say bid, ask, last, mark.. and stored the data?

**Reply by omegastripes** [reply]
Parsed data isn't another string, the purpose of parser is to convert source JSON string to a structure of **VBA arrays** and **dictionary objects** which represent source JSON data. You may use `JSON.ToArray` method to convert single table-like JSON structure into 1d headings array and 2d table data array. First of all you have to extract those tables from the entire parsed data, so that they become separate.

**Reply by omegastripes** [reply]
To make things clear here is the part of the repsonse:
![image](https://user-images.githubusercontent.com/3822668/132067104-8e095d5e-7eb4-41ca-8b68-6cab14296bb8.png)

So, there are lets say Sections, Categories, Subcategories and Properties with some values in the source data. How to show it as a table? Which entities should be converted to rows and which to columns in the resulting table?

**Reply by omegastripes** [reply]
As for me I would split it something like
![image](https://user-images.githubusercontent.com/3822668/132067948-02df8ab4-2d39-4643-9727-919023064aec.png)

It depends on the code you create, but not on parser anyway) And `JSON.ToArray` facilitate to build result.

**Reply by cyclone-isu2001** [reply]
Man - that helps a ton...  so there is a put expiration map and a call expiration map (sections)...  then there are expirations dates (sub-categories), then strike prices (categories), then properties for each strike price (bid, ask last, mark, etc...

a representative table would look like the attached:
![Capture 2](https://user-images.githubusercontent.com/90066957/132068587-d434471e-81ef-45ad-89f3-7d490ed21f22.JPG)

**Reply by omegastripes** [reply]
Source JSON data and representative table have quite different structure, so you should extensively rebuild parts of data to accomplish that. One more to be noted is that actually you displayed not the single table but a list of tables, one table for each date, in this case better to make a single table first on one worksheet, and the create a pivot table to split the data to sections for each date on another worksheet.

**Reply by omegastripes** [reply]
Try something like this

    Sub test()
        Dim resp As String
        Dim data
        Dim state As String
        With CreateObject("MSXML2.XMLHTTP")
            .Open "GET", "<TD Ameritrade API call URL here>", True
            .Send
            Do Until .ReadyState = 4: DoEvents: Loop
            resp = .ResponseText
        End With
        JSON.Parse resp, data, state
        Dim putEdm
        Set putEdm = data("putExpDateMap")
        Dim callEdm
        Set callEdm = data("callExpDateMap")
        Dim resultRows
        Set resultRows = New Dictionary
        Dim keyDate
        For Each keyDate In putEdm
            Debug.Print "Date " & keyDate
            Dim eltDate
            Set eltDate = putEdm(keyDate)
            Dim keyStrike
            For Each keyStrike In eltDate
                Debug.Print "Strike " & keyStrike
                Dim putProps
                Set putProps = eltDate(keyStrike)(0)
                Dim callProps
                Set callProps = callEdm(keyDate)(keyStrike)(0)
                Dim resultRow
                Set resultRow = New Dictionary
                resultRow("exp") = keyDate
                resultRow("strike") = keyStrike
                Set resultRow("put") = putProps
                Set resultRow("call") = callProps
                Set resultRows(resultRows.Count) = resultRow
            Next
        Next
        Dim tableData()
        Dim tableHeader()
        JSON.ToArray resultRows, tableData, tableHeader
        With ThisWorkbook.Sheets(1)
            .Activate
            .Cells.Delete
            With .Cells(1, 1)
                .Resize(1, UBound(tableHeader) - LBound(tableHeader) + 1).Value = tableHeader
                .Offset(1, 0).Resize( _
                        UBound(tableData, 1) - LBound(tableData, 1) + 1, _
                        UBound(tableData, 2) - LBound(tableData, 2) + 1 _
                    ).Value = tableData
            End With
            .Columns.AutoFit
        End With
        MsgBox "Completed"
    End Sub
